causalpy/pymc_models.py

```diff
@@ -507,3 +507,295 @@ def fit(self, X, t, coords):
+            ):
+                self.treatment_effect_param["impulse"] = self.DEFAULT_IMPULSE_PRIOR
+            else:
+                self.treatment_effect_param["impulse"] = self.treatment_effect_param[
+                    "impulse"
+                ]
+
+    def build_model(self, X, y, coords):
+        """
+        Defines the PyMC model
+
+        :param X: An array of the covariates
+        :param y: An array of values representing our outcome y
+        :param coords: Dictionary of named coordinates for PyMC variables (e.g., {"obs_ind": range(n_obs), "coeffs": range(n_covariates)}).
+
+        Assumes the following attributes are already defined in self:
+            - self.timeline: the index of the column in X representing time.
+            - self.time_range: a tuple (lower_bound, upper_bound) for the intervention time.
+            - self.treatment_effect_type: a dictionary specifying which intervention effects to include and their priors.
+        """
+
+        with self:
+            self.add_coords(coords)
+
+            t = pm.Data("t", np.arange(len(X)), dims="obs_ind")
+            X = pm.Data("X", X, dims=["obs_ind", "coeffs"])
+            y = pm.Data("y", y, dims="obs_ind")
+
+            lower_bound = pm.Data("lower_bound", self.time_range[0])
+            upper_bound = pm.Data("upper_bound", self.time_range[1])
+
+            # --- Priors ---
+            treatment_time = pm.Uniform(
+                "treatment_time", lower=lower_bound, upper=upper_bound
+            )
+            beta = pm.Normal(
+                name="beta",
+                mu=self.DEFAULT_BETA_PRIOR[0],
+                sigma=self.DEFAULT_BETA_PRIOR[1],
+                dims="coeffs",
+            )
+
+            # --- Intervention effect ---
+            mu_in_components = []
+
+            if "level" in self.treatment_effect_param:
+                level = pm.Normal(
+                    "level",
+                    mu=self.treatment_effect_param["level"][0],
+                    sigma=self.treatment_effect_param["level"][1],
+                )
+                mu_in_components.append(level)
+            if "trend" in self.treatment_effect_param:
+                trend = pm.Normal(
+                    "trend",
+                    mu=self.treatment_effect_param["trend"][0],
+                    sigma=self.treatment_effect_param["trend"][1],
+                )
+                mu_in_components.append(trend * (t - treatment_time))
+            if "impulse" in self.treatment_effect_param:
+                impulse_amplitude = pm.Normal(
+                    "impulse_amplitude",
+                    mu=self.treatment_effect_param["impulse"][0],
+                    sigma=self.treatment_effect_param["impulse"][1],
+                )
+                decay_rate = pm.HalfNormal(
+                    "decay_rate", sigma=self.treatment_effect_param["impulse"][2]
+                )
+                impulse = pm.Deterministic(
```

Comment:
Nice.

Reply:
Might be worth unpacking the math in words a bit  in  doc-string , for each of the three changepoint models